net/mptcp/mptcp_remp.c

```diff
@@ -0,0 +1,263 @@
+}
+
+/* Corrects the stored skb pointers if they are invalid */
+static void rempsched_correct_skb_pointers(struct sock *meta_sk,
+						struct rempsched_sock_data *sk_data)
+{	
+	struct tcp_sock *meta_tp = tcp_sk(meta_sk);
+	struct mptcp_cb *mpcb = meta_tp->mpcb;
+	struct sk_buff *skb;
+	
+	if (sk_data->send_skb.skb && !after(sk_data->send_skb.skb_end_seq, meta_tp->snd_una))
+		sk_data->send_skb.skb = NULL;
+	
+	if (sk_data->reinject_skb.skb) {
+		if (after(sk_data->send_skb.skb_end_seq, meta_tp->snd_una)) {
+			/* Reinjected packet was not acknowledged yet -> might be still be in reinject queue */
+			if (!skb_queue_empty(&mpcb->reinject_queue)) {
+				skb = skb_peek(&mpcb->reinject_queue);
+				
+				do {
+					if (skb == sk_data->reinject_skb.skb)
+						/* skb is still in the reinjection queue */
+						return;
+				} while (
+					!skb_queue_is_last(&mpcb->reinject_queue, skb) && 
+					(skb = skb_queue_next(&mpcb->reinject_queue, skb)));
+			}
+		}
+		
+		/* Reinject skb is invalid */
+		sk_data->reinject_skb.skb = NULL;
+	}
+}
+
+/* Returns the next skb from the queue */
+static struct sk_buff *remp_next_skb_from_queue(struct sk_buff_head *queue, struct sk_buff *previous)
+{
+	if (skb_queue_empty(queue))
+		return NULL;
+	
+	if (!previous)
+		return skb_peek(queue);
+	
+	if (skb_queue_is_last(queue, previous))
+		return NULL;
+	
+	return skb_queue_next(queue, previous);
+}
+
+static struct sk_buff *remp_next_segment(struct sock *meta_sk,
+					     int *reinject,
+					     struct sock **subsk,
+					     unsigned int *limit)
+{
+	struct tcp_sock *meta_tp = tcp_sk(meta_sk);
+	struct mptcp_cb *mpcb = meta_tp->mpcb;
+	struct rempsched_cb_data *cb_data = rempsched_get_cb_data(meta_tp);
+	struct tcp_sock *first_tp = cb_data->next_subflow;
+	struct tcp_sock *tp;
+	struct rempsched_sock_data *sk_data;
+	struct sk_buff *skb;
+	struct sk_buff *original_skb;
+	
+	if (!first_tp)
+		first_tp = mpcb->connection_list;
+	tp = first_tp;
+	
+	if (skb_queue_empty(&mpcb->reinject_queue) && 
+		skb_queue_empty(&meta_sk->sk_write_queue))
+		/* Nothing to send */
+		return NULL;
+	
+	do {
+		/* Correct the skb pointers of the current subflow */
+		sk_data = rempsched_get_sock_data(tp);
+		rempsched_correct_skb_pointers(meta_sk, sk_data);
+		
+		/* Check if there are skbs that must be sent ... */
+		/* ... in the reinject queue */
+		skb = remp_next_skb_from_queue(&mpcb->reinject_queue, sk_data->reinject_skb.skb);
+		if (skb)
+			*reinject = 1;
+		else {
+			/* ... in the send queue */
+			skb = remp_next_skb_from_queue(&meta_sk->sk_write_queue, sk_data->send_skb.skb);
+			*reinject = 0;
+		}
+		
+		if (skb && mptcp_is_available((struct sock *) tp, skb, false)) {
+			/* Found */
+			cb_data->next_subflow = tp->mptcp->next;
+			
+			if (*reinject) {
+				/* Because the caller will remove the reinject skb from the queue clone it */
+				original_skb = skb;
+				skb = skb_clone(skb, __GFP_NOFAIL);
+				skb_insert(original_skb, skb, &mpcb->reinject_queue);
```

Comment:
When the sending of the reinjected skb fails in mptcp_write_xmit, we end up having original_skb and skb sitting in the reinject-queue.

Overall, I don't think we should send reinjected skb's on all subflows. Because, as they are reinjects, they have already once been sent on all subflows. Thus, I don't think it is necessary to do so.

Or, what use-case do you have in mind?